stream/stream.go

```diff
@@ -0,0 +1,1013 @@
+package stream
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/fxamacker/cbor/v2"
+	"github.com/katzenpost/katzenpost/client"
+	"github.com/katzenpost/katzenpost/core/crypto/rand"
+	"github.com/katzenpost/katzenpost/core/epochtime"
+	"github.com/katzenpost/katzenpost/core/worker"
+	mClient "github.com/katzenpost/katzenpost/map/client"
+	"github.com/katzenpost/katzenpost/map/common"
+	"golang.org/x/crypto/hkdf"
+	"golang.org/x/crypto/nacl/secretbox"
+	"gopkg.in/op/go-logging.v1"
+	"io"
+	"os"
+	"sync"
+	"time"
+)
+
+const (
+	keySize   = 32
+	nonceSize = 24
+)
+
+var (
+	hash               = sha256.New
+	cborFrameOverhead  = 0
+	retryDelay         = epochtime.Period / 16
+	minBackoffDelay    = 1 * time.Millisecond
+	maxBackoffDelay    = 200 * time.Millisecond // epochtime.Period
+	defaultTimeout     = 5 * time.Minute
+	ErrStreamClosed    = errors.New("Stream Closed")
+	ErrFrameDecrypt    = errors.New("Failed to decrypt")
+	ErrGeometryChanged = errors.New("Stream Payload Geometry Change")
+	ErrInvalidAddr     = errors.New("Invalid StreamAddr")
+	ErrHalted          = errors.New("Halted")
+)
+
+// FrameType indicates the state of Stream at the current Frame
+type FrameType uint8
+
+const (
+	// StreamStart indicates that this is the first Frame in a Stream
+	StreamStart FrameType = iota
+	// StreamData indicates that this is a data carrying Frame in a Stream
+	StreamData
+	// StreamEnd indicates that this is the last Frame in a Stream
+	StreamEnd
+)
+
+// Frame is the container for Stream payloads and contains Stream metadata
+// that indicates whether the Frame is the first, last, or an intermediary
+// block. This
+type Frame struct {
```

Comment:
why are some field exported and others not? perhaps all the fields can be unexported?